File: packages/python/src/leetcode/by-problem/0125-valid-palindrome.py
"""
125 - Valid Palindrome
Two Pointers | Middle Convergence | String
Time: O(n) | Space: O(1)

A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all
non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters
and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.
"""


class Solution:
    def isPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s) - 1

        while left < right:
            while not s[left].isalnum() and left < right:
                left += 1
            while not s[right].isalnum() and left < right:
                right -= 1

            if s[left].lower() != s[right].lower():
                return False

            left += 1
            right -= 1

        return True

File: packages/cpp/src/leetcode/by-problem/0125-valid-palindrome.cpp
/*
    125 - Valid Palindrome
    Two pointers | Middle Convergence
    Time: O(n) | Space: O(1)
*/

#include <string>

class Solution
{
public:
    bool isPalindrome(std::string s)
    {
        int l = 0;
        int r = s.size() - 1;

        while (l < r)
        {
            if (s[l] != s[r])
            {
                return false;
            }
            l++;
            r--;
        }

        return true;
    }
};
